File: program/tests/integration-tests/src/assertions.rs
use commerce_program_client::{
    types::Status, Merchant, MerchantOperatorConfig, Operator, Payment, COMMERCE_PROGRAM_ID,
};
use solana_program_pack::Pack;
use solana_sdk::pubkey::Pubkey;
use spl_token::state::Account as TokenAccount;

use crate::utils::{get_token_balance, TestContext};

pub fn assert_operator_account(
    context: &mut TestContext,
    operator_pda: &Pubkey,
    expected_owner: &Pubkey,
    expected_bump: u8,
) {
    let account = context
        .get_account(operator_pda)
        .expect("Operator account should exist");

    assert_eq!(account.owner, COMMERCE_PROGRAM_ID);

    let operator =
        Operator::from_bytes(&account.data).expect("Should deserialize operator account");

    assert_eq!(operator.owner, *expected_owner);
    assert_eq!(operator.bump, expected_bump);
}

pub fn assert_merchant_account(
    context: &mut TestContext,
    merchant_pda: &Pubkey,
    expected_owner: &Pubkey,
    expected_bump: u8,
    expected_settlement_wallet: &Pubkey,
) {
    let merchant_account = context.svm.get_account(merchant_pda).unwrap();
    let merchant_account = Merchant::from_bytes(&merchant_account.data).unwrap();
    assert_eq!(merchant_account.owner, *expected_owner);
    assert_eq!(merchant_account.bump, expected_bump);
    assert_eq!(
        merchant_account.settlement_wallet,
        *expected_settlement_wallet
    );
}

pub fn assert_account_not_exists(context: &mut TestContext, pubkey: &Pubkey) {
    let account = context.get_account(pubkey);
    assert!(account.is_none(), "Account should not exist");
}

pub fn assert_account_exists(context: &mut TestContext, pubkey: &Pubkey) {
    let account = context.get_account(pubkey);
    assert!(account.is_some(), "Account should exist");
}

pub fn assert_account_lamports(context: &mut TestContext, pubkey: &Pubkey, expected_lamports: u64) {
    let account = context.get_account(pubkey).expect("Account should exist");
    assert_eq!(account.lamports, expected_lamports);
}

pub fn assert_account_owner(context: &mut TestContext, pubkey: &Pubkey, expected_owner: &Pubkey) {
    let account = context.get_account(pubkey).expect("Account should exist");
    assert_eq!(account.owner, *expected_owner);
}

pub fn assert_token_account(
    context: &mut TestContext,
    token_account: &Pubkey,
    expected_mint: &Pubkey,
    expected_owner: &Pubkey,
) {
    let account = context
        .get_account(token_account)
        .expect("Account should exist");
    let token_account = TokenAccount::unpack(&account.data).unwrap();
    assert_eq!(token_account.mint, *expected_mint);
    assert_eq!(token_account.owner, *expected_owner);
}

#[allow(clippy::too_many_arguments)]
pub fn assert_merchant_operator_config_account(
    context: &mut TestContext,
    merchant_operator_config_pda: &Pubkey,
    expected_bump: u8,
    expected_version: u32,
    expected_merchant: &Pubkey,
    expected_operator: &Pubkey,
    expected_operator_fee: u64,
    expected_current_order_id: u32,
    _expected_num_policies: u32,
    _expected_num_accepted_currencies: u32,
) {
    let account = context
        .get_account(merchant_operator_config_pda)
        .expect("Merchant operator config account should exist");
    let merchant_operator_config: MerchantOperatorConfig =
        MerchantOperatorConfig::from_bytes(&account.data)
            .expect("Should deserialize merchant operator config account");
    assert_eq!(merchant_operator_config.bump, expected_bump);
    assert_eq!(merchant_operator_config.version, expected_version);
    assert_eq!(merchant_operator_config.merchant, *expected_merchant);
    assert_eq!(merchant_operator_config.operator, *expected_operator);
    assert_eq!(merchant_operator_config.operator_fee, expected_operator_fee);
    assert_eq!(
        merchant_operator_config.current_order_id,
        expected_current_order_id
    );
    // todo validate policies and accepted currencies
}

pub fn assert_payment_account(
    context: &mut TestContext,
    payment_pda: &Pubkey,
    expected_order_id: u32,
    expected_amount: u64,
    expected_status: Status,
) {
    let account = context
        .get_account(payment_pda)
        .expect("Payment account should exist");

    assert_eq!(account.owner, COMMERCE_PROGRAM_ID);

    let payment = Payment::from_bytes(&account.data).expect("Should deserialize payment account");

    assert_eq!(payment.order_id, expected_order_id);
    assert_eq!(payment.amount, expected_amount);
    assert_eq!(payment.status, expected_status);
    assert!(payment.created_at > 0, "Created timestamp should be set");
}

pub fn assert_token_balance_changes(
    context: &mut TestContext,
    pre_balances: [u64; 2],
    sender_ata: &Pubkey,
    receiver_ata: &Pubkey,
    expected_amount: u64,
) {
    let sender_post_balance = get_token_balance(context, sender_ata);
    let receiver_post_balance = get_token_balance(context, receiver_ata);

    let sender_pre_balance = pre_balances[0];
    let receiver_pre_balance = pre_balances[1];

    // Assert sender balance decreased by expected amount
    assert_eq!(
        sender_pre_balance - expected_amount,
        sender_post_balance,
        "Sender balance should decrease by {} (pre: {}, post: {})",
        expected_amount,
        sender_pre_balance,
        sender_post_balance
    );

    // Assert receiver balance increased by expected amount
    assert_eq!(
        receiver_pre_balance + expected_amount,
        receiver_post_balance,
        "Receiver balance should increase by {} (pre: {}, post: {})",
        expected_amount,
        receiver_pre_balance,
        receiver_post_balance
    );
}

pub struct BalanceChange {
    pub ata: Pubkey,
    pub expected_change: i64, // Positive for increase, negative for decrease
    pub description: String,
}

pub fn assert_multiple_token_balance_changes(
    context: &mut TestContext,
    pre_balances: &[(Pubkey, u64)], // Vec of (ata, pre_balance)
    balance_changes: &[BalanceChange],
) {
    // Get all post balances
    let post_balances: Vec<(Pubkey, u64)> = balance_changes
        .iter()
        .map(|change| (change.ata, get_token_balance(context, &change.ata)))
        .collect();

    // Assert each balance change
    for (i, change) in balance_changes.iter().enumerate() {
        let pre_balance = pre_balances
            .iter()
            .find(|(ata, _)| *ata == change.ata)
            .map(|(_, balance)| *balance)
            .expect("Pre-balance should exist for ATA");

        let post_balance = post_balances[i].1;
        let expected_post_balance = if change.expected_change >= 0 {
            pre_balance + (change.expected_change as u64)
        } else {
            pre_balance - ((-change.expected_change) as u64)
        };

        assert_eq!(
            expected_post_balance,
            post_balance,
            "{} (expected change: {}, pre: {}, post: {})",
            change.description,
            change.expected_change,
            pre_balance,
            post_balance
        );
    }
}

File: program/clients/typescript/tests/integration/helpers/assertions.ts
import { Address, getBase64Codec, SolanaClient } from 'gill';
import { fetchToken } from 'gill/programs';
import { expect } from '@jest/globals';
import {
    fetchMerchant,
    fetchOperator,
    fetchMerchantOperatorConfig,
    Status,
    fetchPayment,
    PolicyData,
} from '../../../src/generated';
import { getMerchantOperatorConfigCodec } from '../../../src/codecs/merchantOperatorConfig';

export async function assertMerchantAccount({
    client,
    merchantPda,
    expectedOwner,
    expectedBump,
    expectedSettlementWallet
}: {
    client: SolanaClient,
    merchantPda: Address,
    expectedOwner: Address,
    expectedBump: number,
    expectedSettlementWallet: Address
}
) {
    const merchantAccount = await fetchMerchant(client.rpc, merchantPda, { commitment: 'processed' });
    expect(merchantAccount.data).not.toBeNull();
    expect(merchantAccount.data.owner).toBe(expectedOwner);
    expect(merchantAccount.data.bump).toBe(expectedBump);
    expect(merchantAccount.data.settlementWallet).toBe(expectedSettlementWallet);
}

export async function assertOperatorAccount({
    client,
    operatorPda,
    expectedOwner,
    expectedBump
}: {
    client: SolanaClient,
    operatorPda: Address,
    expectedOwner: Address,
    expectedBump: number
}) {
    const operatorAccount = await fetchOperator(client.rpc, operatorPda, { commitment: 'processed' });
    expect(operatorAccount.data).not.toBeNull();
    expect(operatorAccount.data.owner).toBe(expectedOwner);
    expect(operatorAccount.data.bump).toBe(expectedBump);
}

export async function assertTokenAccount({
    client,
    tokenAccount,
    expectedMint,
    expectedOwner
}: {
    client: SolanaClient,
    tokenAccount: Address,
    expectedMint: Address,
    expectedOwner: Address
}) {
    const accountInfo = await fetchToken(client.rpc, tokenAccount, { commitment: 'processed' });
    expect(accountInfo.data).not.toBeNull();
    expect(accountInfo.data.mint).toBe(expectedMint);
    expect(accountInfo.data.owner).toBe(expectedOwner);
}

export async function assertMerchantOperatorConfigAccount({
    client,
    merchantOperatorConfigPda,
    expectedBump,
    expectedVersion,
    expectedMerchant,
    expectedOperator,
    expectedOperatorFee,
    expectedCurrentOrderId,
    expectedPolicies,
    expectedAcceptedCurrencies
}: {
    client: SolanaClient,
    merchantOperatorConfigPda: Address,
    expectedBump: number,
    expectedVersion: number,
    expectedMerchant: Address,
    expectedOperator: Address,
    expectedOperatorFee: bigint,
    expectedCurrentOrderId: number,
    expectedPolicies: PolicyData[],
    expectedAcceptedCurrencies: Address[]
}) {
    //const merchantOperatorConfig = await fetchMerchantOperatorConfig(client.rpc, merchantOperatorConfigPda, { commitment: 'processed' });
    const safeLogBigint = (value: any) => {
        return JSON.stringify(value, (key, value) => {
            if (typeof value === 'bigint') {
                return value.toString();
            }
            return value;
        });
    }
    const base64 = await client.rpc.getAccountInfo(merchantOperatorConfigPda, { commitment: 'processed', encoding: 'base64' }).send();
    expect(base64.value?.data[0]).not.toBeNull();
    const data = base64.value?.data[0] as string;
    const tranform = getBase64Codec().encode(data);
    const merchantOperatorConfig = getMerchantOperatorConfigCodec().decode(tranform);

    expect(merchantOperatorConfig).not.toBeNull();
    expect(merchantOperatorConfig.bump).toBe(expectedBump);
    expect(merchantOperatorConfig.version).toBe(expectedVersion);
    expect(merchantOperatorConfig.merchant).toBe(expectedMerchant);
    expect(merchantOperatorConfig.operator).toBe(expectedOperator);
    expect(merchantOperatorConfig.operatorFee).toBe(expectedOperatorFee);
    expect(merchantOperatorConfig.currentOrderId).toBe(expectedCurrentOrderId);
    expect(merchantOperatorConfig.numPolicies).toBe(expectedPolicies.length);
    expect(merchantOperatorConfig.numAcceptedCurrencies).toBe(expectedAcceptedCurrencies.length);
    expect(merchantOperatorConfig.acceptedCurrencies).toEqual(expectedAcceptedCurrencies);
    expect(merchantOperatorConfig.policies.length).toBe(expectedPolicies.length);
    
    merchantOperatorConfig.policies.forEach((policy, index) => {
        const expectedPolicy = expectedPolicies[index];
        expect(policy.__kind).toBe(expectedPolicy.__kind);
        
        // Compare fields based on policy type
        if (policy.__kind === 'Settlement' && expectedPolicy.__kind === 'Settlement') {
            const actualData = policy.fields[0];
            const expectedData = expectedPolicy.fields[0];
            
            expect(actualData.minSettlementAmount).toBe(BigInt(expectedData.minSettlementAmount));
            expect(actualData.settlementFrequencyHours).toBe(expectedData.settlementFrequencyHours);
            expect(actualData.autoSettle).toBe(expectedData.autoSettle);
        } else if (policy.__kind === 'Refund' && expectedPolicy.__kind === 'Refund') {
            const actualData = policy.fields[0];
            const expectedData = expectedPolicy.fields[0];
            
            expect(actualData.maxAmount).toBe(BigInt(expectedData.maxAmount));
            expect(actualData.maxTimeAfterPurchase).toBe(BigInt(expectedData.maxTimeAfterPurchase));
        }
    });
}


export async function assertPaymentAccount({
    client,
    paymentPda,
    expectedOrderId,
    expectedAmount,
    expectedStatus,
}: {
    client: SolanaClient,
    paymentPda: Address,
    expectedOrderId: number,
    expectedAmount: bigint,
    expectedStatus: Status,
}) {
    const paymentAccount = await fetchPayment(client.rpc, paymentPda, { commitment: 'processed' });
    expect(paymentAccount.data.orderId).toBe(expectedOrderId);
    expect(paymentAccount.data.amount).toBe(expectedAmount);
    expect(paymentAccount.data.status).toBe(expectedStatus);
    expect(paymentAccount.data.createdAt).toBeDefined();
    expect(paymentAccount.data.discriminator).toBe(3);
}

// Helper function to get token balance
export async function getTokenBalance(client: SolanaClient, tokenAccount: Address): Promise<bigint> {
    const balance = await client.rpc.getTokenAccountBalance(tokenAccount, { commitment: 'processed' }).send();
    return BigInt(balance.value.amount);
}

// Assert token balance changes for simple two-party transfers
export async function assertTokenBalanceChanges({
    client,
    preBalances,
    senderAta,
    receiverAta,
    expectedAmount,
}: {
    client: SolanaClient,
    preBalances: [bigint, bigint], // [senderPreBalance, receiverPreBalance]
    senderAta: Address,
    receiverAta: Address,
    expectedAmount: bigint,
}) {
    const senderPostBalance = await getTokenBalance(client, senderAta);
    const receiverPostBalance = await getTokenBalance(client, receiverAta);

    const [senderPreBalance, receiverPreBalance] = preBalances;

    // Assert sender balance decreased by expected amount
    expect(senderPostBalance).toBe(senderPreBalance - expectedAmount);

    // Assert receiver balance increased by expected amount
    expect(receiverPostBalance).toBe(receiverPreBalance + expectedAmount);
}

// Balance change interface for complex multi-party transfers
export interface BalanceChange {
    ata: Address;
    expectedChange: bigint; // Positive for increase, negative for decrease
    description: string;
}

// Assert multiple token balance changes for complex transfers
export async function assertMultipleTokenBalanceChanges({
    client,
    preBalances,
    balanceChanges,
}: {
    client: SolanaClient,
    preBalances: Map<Address, bigint>, // Map of ATA to pre-balance
    balanceChanges: BalanceChange[],
}) {
    // Get all post balances and assert each change
    for (const change of balanceChanges) {
        const preBalance = preBalances.get(change.ata);
        if (preBalance === undefined) {
            throw new Error(`Pre-balance not found for ATA: ${change.ata}`);
        }

        const postBalance = await getTokenBalance(client, change.ata);
        const expectedPostBalance = preBalance + change.expectedChange;

        expect(postBalance).toBe(expectedPostBalance);
    }
}